### PR TITLE
Single-kernel GEMM + layernorm

### PR DESCRIPTION
**Pull request by rosenrodt**
The GEMM + Layernorm implementation is a specialized kernel which allows fusing both layers together given the condition that GEMM extent N of MNK is spanned by a single workgroup. For example, a kernel configured with NPerBlock = 128 is allowed to operate on GEMM sizes with N <= 128

Full data path: GEMM -> bias -> activation -> add -> layernorm

Diff between gemm_xdl_cshuffle and gemm_xdl_layernorm_cshuffle
[include/ck/tensor_operation/gpu/device/device_gemm_xdl_layernorm_cshuffle.hpp](https://github.com/ROCmSoftwarePlatform/composable_kernel/compare/83fde45...ROCmSoftwarePlatform:rosenrodt/gemm-layernorm#diff-8ba88325f882f4c3d5beba324850d753dc27ddee74de5ff0e906f9e6e15c3f06)
[include/ck/tensor_operation/gpu/grid/gridwise_gemm_xdl_layernorm_cshuffle_v1.hpp](https://github.com/ROCmSoftwarePlatform/composable_kernel/compare/83fde45...ROCmSoftwarePlatform:rosenrodt/gemm-layernorm#diff-20f072a44a58d0fe00959a6e5eafab6bd5e4e6238c28480d4bc3fb897dcce710)

Ongoing
- [x] Host ReferenceGemmLayernorm implementation & example code
- [x] Device kernel implementation
- [x] Update C tile map
- [x] Combine reduction LDS workspace with singly allocated LDS memory used for blockwise copy and c-shuffle
- [x] C0 (bias/add/gamma/beta) has its own precision type  

Future action
- [ ] (int8) Type of p_c0_bias and p_c0_add should be different https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/263#discussion_r892112579
- (hold) ~Split gridwise gemm into separate GEMM and epilogue (layernorm in this case) to allow code reuse~
- (hold) ~Dedicated test for different param & size combos~
- (hold) ~Performance~

### CODE REVIEW

File: example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp
Comment (rocking5566) [inline example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp:249]: Please remove debug code

File: example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp
Comment (rocking5566) [inline example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp:241]: Just to make sure. Should we return the result of gemm or only the result of layerNorm?

File: example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp
Comment (rocking5566) [inline example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp:126]: Just to make sure. 
Should we return the both of the result of gemm or just the result of layerNorm?
If we cannot the make sure the answer, I suggest to pass the global mem of gemm result in the parameter and check whether to return by the nuulptr.

File: example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp
Comment (rocking5566) [inline example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp:40]: For flexible, I suggest to separate LayerNormElementOp from CElementOp. (4 elementwiseOP in total)

File: example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp
Comment (rosenrodt) [reply example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp:126]: The result should return the end result of gemm + layernorm. Cc @asroy

File: example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp
Comment (rosenrodt) [reply example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp:40]: Can you make an example for the use case? What is the input and output of LayernormElementwiseOp in this context?

File: include/ck/tensor_operation/gpu/device/device_gemm_xdl_layernorm_cshuffle.hpp
Comment (rocking5566) [inline include/ck/tensor_operation/gpu/device/device_gemm_xdl_layernorm_cshuffle.hpp:35]: Add LayerNorm Elementwise operation

File: include/ck/tensor_operation/gpu/device/device_gemm_xdl_layernorm_cshuffle.hpp
Comment (rocking5566) [inline include/ck/tensor_operation/gpu/device/device_gemm_xdl_layernorm_cshuffle.hpp:636]: How about tuple of bias? This might be next PR.

File: example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp
Comment (rocking5566) [inline example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp:31]: How about adding bias, gamma and beta type?
Especially in int8 implementation, bias would be same as acc type.

File: example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp
Comment (rosenrodt) [reply example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp:40]: I got it. Fair point

File: example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp
Comment (rosenrodt) [reply example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp]: I have added extra precision type for C0 which applies to bias/gamma/beta.

File: include/ck/tensor_operation/gpu/device/device_gemm_xdl_layernorm_cshuffle.hpp
Comment (rosenrodt) [reply include/ck/tensor_operation/gpu/device/device_gemm_xdl_layernorm_cshuffle.hpp:38]: Addressed in [AccElemOp for gemm outputs prior to feeding to layernorm](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/263/commits/ddc76a8bc56c4445ba808f43e85008dc49d662a1)

**Comment by rosenrodt** [comment]
WIP/draft status removed and ready to review

File: example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp
Comment (rocking5566) [reply example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp]: Because in this class, you already have CLayout, CDataType, c_grid_desc,  CShuffleXXXXX, CReduceThreadXXXX, block_2_ctile_map...and so on.
Hence, I thought naming of CElementwiseOperation and p_c_grid_ (global mem) might be confusing.

File: example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp
Comment (rocking5566) [inline example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp:54]: I suggest not to name c after output of layerNorm.
This might confuse with CLayout

File: include/ck/tensor_operation/gpu/grid/gridwise_gemm_xdl_layernorm_cshuffle_v1.hpp
Comment (rocking5566) [inline include/ck/tensor_operation/gpu/grid/gridwise_gemm_xdl_layernorm_cshuffle_v1.hpp:742]: Use FloatReduceAcc for c0_thread_buf?

File: example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp
Comment (asroy) [reply example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp:142]: yes, only the result of layernorm

File: include/ck/tensor_operation/gpu/grid/gridwise_gemm_xdl_layernorm_cshuffle_v1.hpp
Comment (rosenrodt) [reply include/ck/tensor_operation/gpu/grid/gridwise_gemm_xdl_layernorm_cshuffle_v1.hpp:742]: It is loaded as FloatC0 type but will be cast to FloatReduceAcc upon doing computation.

Comment (rocking5566) [reply include/ck/tensor_operation/gpu/grid/gridwise_gemm_xdl_layernorm_cshuffle_v1.hpp:742]: Use FloatReduceAcc to prevent implicit cast in the following
 `c_reduce_thread_buf(i) += c0_thread_buf(i);`

Comment (rosenrodt) [reply include/ck/tensor_operation/gpu/grid/gridwise_gemm_xdl_layernorm_cshuffle_v1.hpp:749]: Added [explicit cast](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/263/commits/31b3f1dcb2a17907f3530c2d8281c1a666cf7185)

File: include/ck/tensor_operation/gpu/device/device_gemm_xdl_layernorm_cshuffle.hpp
Comment (rosenrodt) [reply include/ck/tensor_operation/gpu/device/device_gemm_xdl_layernorm_cshuffle.hpp]: Not required as of now

File: example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp
Comment (rocking5566) [inline example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp:53]: The context of acc has two different meanings here. One for gemm acc, the other for reduceAcc.
How about GemmAccElementOp?

File: include/ck/tensor_operation/gpu/device/device_gemm_xdl_layernorm_cshuffle.hpp
Comment (rocking5566) [inline include/ck/tensor_operation/gpu/device/device_gemm_xdl_layernorm_cshuffle.hpp:665]: Type of p_c0_bias and p_c0_add should be different.
In the use case of int8 quantization. p_a, p_b, p_c, p_c0_add would be int8.
whereas p_c0_bias would be int32.

File: example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp
Comment (rocking5566) [inline example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp:50]: In the use case of quantization, I thought we need functor for add term
image

Comment (rosenrodt) [reply example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp:50]: INT8 is not required by task https://ontrack-internal.amd.com/browse/LWPCK-256. We can revisit this in later tasks.

**Comment by rosenrodt** [comment]
This PR needs clarification for defining the interface, and ultimately should be consistent with #272

It also conflicts on changes made in #274, so I think we put this PR on hold until #272 #274 are merged and I will revise this PR accordingly.

**Review by rocking5566 — APPROVED** [review]

File: example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp
Comment (rocking5566) [inline example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp:5]: Remove duplicated header

**Review by rocking5566 — CHANGES_REQUESTED** [review]

File: example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp
Comment (rosenrodt) [reply example/21_gemm_layernorm/gemm_xdl_layernorm_single_kernel_fp16.cpp]: extra header removed

**Review by rocking5566 — DISMISSED** [review]

**Comment by zjing14** [comment]
@rocking5566 Please resolve conflicts.

**Comment by rosenrodt** [comment]
> @rocking5566 Please resolve conflicts.

@zjing14 I plan to hold until #281 is merged

File: include/ck/tensor_operation/gpu/thread/reduction_functions_threadwise.hpp
Comment (rocking5566) [inline include/ck/tensor_operation/gpu/thread/reduction_functions_threadwise.hpp:33]: Why you add this?

**Review by rocking5566 — CHANGES_REQUESTED** [review]

**Review by rocking5566 — APPROVED** [review]

**Review by asroy — APPROVED** [review]

**Pull request merged**
